AI tests broken on Python 3.11

Bug Report
==========
<!--
Please fill in a a meaningful and concise title in the field above.
It helps us to classify the issue even  before reading the body and handling the
issue a bit better.

Examples:
  * Great: "Unable to select fleet in galaxy map"
  * Good: "Unable to click fleet"
  * Not really okay: "Issue with fleets"
  * Horrible: "HELP!!!"
-->


Environment
-----------
<!--
You don't need to provide the information in section when they are not useful in
the context of this issue.  For example a crashing hosted server may not need
information about the graphic card, but maybe depend on the operating system you
use.  When in doubt provide all information you have available.

You can find the FreeOrion version number in the lower right corner of the game
main screen.

When stating the operating system also note the version of the operating system
so, `Windows 8 Pro` instead of `Windows` or `Mac OSX Mountain Lion` instead of
`OSX`.
-->

* **FreeOrion Version**: master branch
* **Operating System**: Linux Ubuntu
* **Fetched as** <!-- delete lines below that don't apply -->
  * Compiled from source


Description
-----------
<!--
Add a meaningful description of the bug you encountered.

If a screenshot or image helps to describe the issue content feel free to create
one.

Also attach the log files the game creates.  The log files The log files can be
found on
* Windows: %APPDATA%\FreeOrion
* MacOSX: ~/Library/"Application Support"/FreeOrion
* Linux: ${XDG_DATA_HOME:-~/.local/share}/freeorion
-->

https://github.com/freeorion/freeorion/actions/runs/15223649284/job/42822950992?pr=5234

```
=================================== FAILURES ===================================
______________________________ test_setstate_call ______________________________

o = <test_savegame_manager.SetStateTester object at 0x7f0c6dc417d0>

    def encode(o: Any) -> str:
        """Encode the passed object as json-based string.
    
        :param o: object to be encoded
        :return: String representation of the object state
        """
        o_type = type(o)
    
        # Find and call the correct encoder based
        # on the type of the object to encode
        try:
>           encoder = _encoder_table[o_type]
E           KeyError: <class 'test_savegame_manager.SetStateTester'>

o          = <test_savegame_manager.SetStateTester object at 0x7f0c6dc417d0>
o_type     = <class 'test_savegame_manager.SetStateTester'>

default/python/AI/savegame_codec/_encoder.py:70: KeyError

During handling of the above exception, another exception occurred:

    @pytest.mark.usefixtures("trusted_scope")
    def test_setstate_call():
        with pytest.raises(Success):
>           savegame_codec.decode(savegame_codec.encode(SetStateTester()))


default/python/tests/AI/save_game_codec/test_savegame_manager.py:168: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
default/python/AI/savegame_codec/_encoder.py:75: in encode
    return _encode_object(o)
        o          = <test_savegame_manager.SetStateTester object at 0x7f0c6dc417d0>
        o_type     = <class 'test_savegame_manager.SetStateTester'>
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = <test_savegame_manager.SetStateTester object at 0x7f0c6dc417d0>

    def _encode_object(obj):
        """Get a string representation of state of an object which is not handled by a specialized encoder."""
        try:
            class_name = f"{obj.__class__.__module__}.{obj.__class__.__name__}"
            if class_name not in trusted_classes:
                raise CanNotSaveGameException("Class %s is not trusted" % class_name)
        except AttributeError:
            # obj does not have a class or class has no module
            raise CanNotSaveGameException(f"Encountered unsupported object {obj} ({type(obj)})")
    
        # if possible, use getstate method to query the state, otherwise use the object's __dict__
        try:
            getstate = obj.__getstate__
        except AttributeError:
            value = obj.__dict__
        else:
            # only call now to avoid catching exceptions raised during the getstate call
            value = getstate()
    
        # encode information about class
>       value.update({"__class__": obj.__class__.__name__, "__module__": obj.__class__.__module__})
E       AttributeError: 'NoneType' object has no attribute 'update'

class_name = 'test_savegame_manager.SetStateTester'
getstate   = <built-in method __getstate__ of SetStateTester object at 0x7f0c6dc417d0>
obj        = <test_savegame_manager.SetStateTester object at 0x7f0c6dc417d0>
value      = None

default/python/AI/savegame_codec/_encoder.py:129: AttributeError
=========================== short test summary info ============================
FAILED default/python/tests/AI/save_game_codec/test_savegame_manager.py::test_setstate_call - AttributeError: 'NoneType' object has no attribute 'update'
======================== 1 failed, 212 passed in 0.67s =========================
```


Expected Result
---------------
<!--
When you enter an issue please add a description of what behaviour you would
expect from the game instead of the issue.
-->

Tests are passed

Steps to reproduce
------------------
<!--
If the error only manifests itself after doing a certain number of actions
please add a list of steps to reproduce the bug.  If the error only occurs in
a certain game state please be sure to add a save game to the issue.

To upload save games please pack them into a Zip archive (*.zip) first.
-->
* Install Python 3.11
* Launch pytests

--- a/.github/workflows/_lint-python.yml
+++ b/.github/workflows/_lint-python.yml
@@ -20,7 +20,11 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ['3.10'] # , '3.11', '3.12', '3.13']
+        python-version:
+          - '3.10'
+          - '3.11'
+          - '3.12'
+          - '3.13'
     steps:
       - uses: actions/checkout@v4
       - name: Set up Python ${{ matrix.python-version }}

--- a/default/python/AI/savegame_codec/_decoder.py
+++ b/default/python/AI/savegame_codec/_decoder.py
@@ -29,7 +29,7 @@ from ._definitions import (
     TRUE,
     TUPLE_PREFIX,
     InvalidSaveGameException,
-    trusted_classes,
+    TrustedClasses,
 )
 
 
@@ -58,6 +58,7 @@ def load_savegame_string(string: str | bytes) -> AIstate:
     import base64
     import zlib
 
+    trusted_classes = TrustedClasses()
     try:
         new_string = base64.b64decode(string)
     except (binascii.Error, ValueError, TypeError) as e:
@@ -66,17 +67,18 @@ def load_savegame_string(string: str | bytes) -> AIstate:
         new_string = zlib.decompress(new_string)
     except zlib.error as e:
         raise SaveDecompressException("Fail to decompress savestate %s" % e) from e
-    return decode(new_string.decode("utf-8"))
+    return decode(new_string.decode("utf-8"), trusted_classes)
 
 
-def decode(obj):
-    return _FreeOrionAISaveGameDecoder().decode(obj)
+def decode(obj, trusted_classes: TrustedClasses):
+    return _FreeOrionAISaveGameDecoder(trusted_classes).decode(obj)
 
 
 class _FreeOrionAISaveGameDecoder(json.JSONDecoder):
-    def __init__(self, **kwargs):
+    def __init__(self, trusted_classes: TrustedClasses):
         # do not allow control characters
-        super().__init__(strict=True, **kwargs)
+        self._trusted_classes = trusted_classes
+        super().__init__(strict=True)
 
     def decode(self, s, _w=None):
         # use the default JSONDecoder to parse the string into a dict
@@ -94,7 +96,7 @@ class _FreeOrionAISaveGameDecoder(json.JSONDecoder):
         class_name = obj.pop("__class__")
         module_name = obj.pop("__module__")
         full_name = f"{module_name}.{class_name}"
-        cls = trusted_classes.get(full_name)
+        cls = self._trusted_classes.get(full_name)
         if cls is None:
             raise InvalidSaveGameException("DANGER DANGER - %s not trusted" % full_name)
 

--- a/default/python/AI/savegame_codec/_definitions.py
+++ b/default/python/AI/savegame_codec/_definitions.py
@@ -6,30 +6,42 @@ import ColonisationAI
 import CombatRatingsAI
 import fleet_orders
 
-trusted_classes = {
-    f"{cls.__module__}.{cls.__name__}": cls
-    for cls in [
-        AIFleetMission.AIFleetMission,
-        fleet_orders.AIFleetOrder,
-        fleet_orders.OrderMilitary,
-        fleet_orders.OrderDefend,
-        fleet_orders.OrderColonize,
-        fleet_orders.OrderOutpost,
-        fleet_orders.OrderPause,
-        fleet_orders.OrderInvade,
-        fleet_orders.OrderMove,
-        fleet_orders.OrderRepair,
-        fleet_orders.OrderResupply,
-        character.character_module.Trait,
-        character.character_module.Aggression,
-        character.character_module.EmpireIDTrait,
-        character.character_module.Character,
-        AIstate.AIstate,
-        ColonisationAI.OrbitalColonizationManager,
-        ColonisationAI.OrbitalColonizationPlan,
-        CombatRatingsAI.ShipCombatStats,
-    ]
-}
+
+class TrustedClasses:
+    def __init__(self, extra_classes: dict | None = None):
+        self._trusted_classes = {
+            f"{cls.__module__}.{cls.__name__}": cls
+            for cls in [
+                AIFleetMission.AIFleetMission,
+                fleet_orders.AIFleetOrder,
+                fleet_orders.OrderMilitary,
+                fleet_orders.OrderDefend,
+                fleet_orders.OrderColonize,
+                fleet_orders.OrderOutpost,
+                fleet_orders.OrderPause,
+                fleet_orders.OrderInvade,
+                fleet_orders.OrderMove,
+                fleet_orders.OrderRepair,
+                fleet_orders.OrderResupply,
+                character.character_module.Trait,
+                character.character_module.Aggression,
+                character.character_module.EmpireIDTrait,
+                character.character_module.Character,
+                AIstate.AIstate,
+                ColonisationAI.OrbitalColonizationManager,
+                ColonisationAI.OrbitalColonizationPlan,
+                CombatRatingsAI.ShipCombatStats,
+            ]
+        }
+        if extra_classes:
+            self._trusted_classes.update(extra_classes)
+
+    def __contains__(self, name: str):
+        return name in self._trusted_classes
+
+    def get(self, name: str):
+        return self._trusted_classes.get(name)
+
 
 # prefixes to encode types not supported by json
 # or not fully supported as dictionary key

--- a/default/python/AI/savegame_codec/_encoder.py
+++ b/default/python/AI/savegame_codec/_encoder.py
@@ -21,8 +21,9 @@ import base64
 import collections
 import json
 import zlib
+from collections.abc import Callable
 from enum import IntEnum
-from typing import Any, Callable
+from typing import Any
 
 from ._definitions import (
     ENUM_PREFIX,

--- a/default/python/AI/savegame_codec/_encoder.py
+++ b/default/python/AI/savegame_codec/_encoder.py
@@ -22,7 +22,7 @@ import collections
 import json
 import zlib
 from enum import IntEnum
-from typing import Any
+from typing import Any, Callable
 
 from ._definitions import (
     ENUM_PREFIX,
@@ -35,7 +35,7 @@ from ._definitions import (
     TRUE,
     TUPLE_PREFIX,
     CanNotSaveGameException,
-    trusted_classes,
+    TrustedClasses,
 )
 
 
@@ -52,11 +52,12 @@ def build_savegame_string() -> bytes:
     """
     from aistate_interface import get_aistate
 
-    savegame_string = encode(get_aistate())
+    trusted_classes = TrustedClasses()
+    savegame_string = encode(get_aistate(), trusted_classes=trusted_classes)
     return base64.b64encode(zlib.compress(savegame_string.encode("utf-8")))
 
 
-def encode(o: Any) -> str:
+def encode(o: Any, trusted_classes: TrustedClasses) -> str:
     """Encode the passed object as json-based string.
 
     :param o: object to be encoded
@@ -66,99 +67,98 @@ def encode(o: Any) -> str:
 
     # Find and call the correct encoder based
     # on the type of the object to encode
+
+    type_encoder = TypeEncoder(trusted_classes)
     try:
-        encoder = _encoder_table[o_type]
+        encoder_function = type_encoder.get_encoder_function(o_type)
     except KeyError:
         if issubclass(o_type, IntEnum):
             return _encode_with_prefix(ENUM_PREFIX, f"{o.__class__.__name__}.{o.name}")
         else:
-            return _encode_object(o)
+            return type_encoder._encode_object(o)
     else:
         # only call now to not catch KeyError withing the encoder call
-        return encoder(o)
+        return encoder_function(o)
 
 
-def _encode_str(o):
-    """
-    Encode string.
+class TypeEncoder:
+    def __init__(self, trusted_classes: TrustedClasses):
+        self._trusted_classes = trusted_classes
 
-    We don't check if prefixes for types (eg __ENUM__) is in the string.
-    The string will be encoded, but it will not be decoded back.
-    Since encoded object is under control of developers team,
-     and it is possible to fix it manually by editing save game, we will leave it as is.
-    """
-    return json.dumps(o)
+    def _encode_str(self, o):
+        """
+        Encode string.
 
+        We don't check if prefixes for types (eg __ENUM__) is in the string.
+        The string will be encoded, but it will not be decoded back.
+        Since encoded object is under control of developers team,
+         and it is possible to fix it manually by editing save game, we will leave it as is.
+        """
+        return json.dumps(o)
 
-def _encode_none(o):
-    return _encode_with_prefix(NONE, "")
+    def _encode_dict(self, o):
+        """Get a string representation of a dict with its encoded content."""
+        return "{%s}" % (
+            ", ".join([f"{encode(k, self._trusted_classes)}: {encode(v, self._trusted_classes)}" for k, v in o.items()])
+        )
 
+    def _encode_set(self, o):
+        """Get a string representation of a set with its encoded content."""
+        return _encode_with_prefix(SET_PREFIX, "(%s)" % (self._encode_list(list(o)).replace('"', PLACEHOLDER)))
 
-def _encode_int(o):
-    return _encode_with_prefix(INT_PREFIX, str(o))
+    def _encode_list(self, o):
+        """Get a string representation of a list with its encoded content."""
+        return "[%s]" % (", ".join([encode(v, self._trusted_classes) for v in o]))
 
+    def _encode_tuple(self, o):
+        """Get a string representation of a tuple with its encoded content."""
+        return _encode_with_prefix(TUPLE_PREFIX, "(%s)" % (self._encode_list(list(o)).replace('"', PLACEHOLDER)))
 
-def _encode_float(o):
-    return _encode_with_prefix(FLOAT_PREFIX, repr(o))
+    def _encode_none(self, o):
+        return _encode_with_prefix(NONE, "")
 
+    def _encode_int(self, o):
+        return _encode_with_prefix(INT_PREFIX, str(o))
 
-def _encode_bool(o):
-    return _encode_with_prefix(TRUE, "") if o else _encode_with_prefix(FALSE, "")
+    def _encode_float(self, o):
+        return _encode_with_prefix(FLOAT_PREFIX, repr(o))
 
+    def _encode_bool(self, o):
+        return _encode_with_prefix(TRUE, "") if o else _encode_with_prefix(FALSE, "")
 
-def _encode_object(obj):
-    """Get a string representation of state of an object which is not handled by a specialized encoder."""
-    try:
-        class_name = f"{obj.__class__.__module__}.{obj.__class__.__name__}"
-        if class_name not in trusted_classes:
-            raise CanNotSaveGameException("Class %s is not trusted" % class_name)
-    except AttributeError:
-        # obj does not have a class or class has no module
-        raise CanNotSaveGameException(f"Encountered unsupported object {obj} ({type(obj)})")
+    def _encode_object(self, obj):
+        """Get a string representation of state of an object which is not handled by a specialized encoder."""
+        try:
+            class_name = f"{obj.__class__.__module__}.{obj.__class__.__name__}"
+            if class_name not in self._trusted_classes:
+                raise CanNotSaveGameException("Class %s is not trusted" % class_name)
+        except AttributeError:
+            # obj does not have a class or class has no module
+            raise CanNotSaveGameException(f"Encountered unsupported object {obj} ({type(obj)})")
 
-    # if possible, use getstate method to query the state, otherwise use the object's __dict__
-    try:
-        getstate = obj.__getstate__
-    except AttributeError:
-        value = obj.__dict__
-    else:
-        # only call now to avoid catching exceptions raised during the getstate call
-        value = getstate()
+        # if possible, use getstate method to query the state, otherwise use the object's __dict__
+        try:
+            getstate = obj.__getstate__
+        except AttributeError:
+            value = obj.__dict__
+        else:
+            # only call now to avoid catching exceptions raised during the getstate call
+            value = getstate()
 
-    # encode information about class
-    value.update({"__class__": obj.__class__.__name__, "__module__": obj.__class__.__module__})
-    return _encode_dict(value)
+        # encode information about class
+        value.update({"__class__": obj.__class__.__name__, "__module__": obj.__class__.__module__})
+        return self._encode_dict(value)
 
-
-def _encode_list(o):
-    """Get a string representation of a list with its encoded content."""
-    return "[%s]" % (", ".join([encode(v) for v in o]))
-
-
-def _encode_tuple(o):
-    """Get a string representation of a tuple with its encoded content."""
-    return _encode_with_prefix(TUPLE_PREFIX, "(%s)" % (_encode_list(list(o)).replace('"', PLACEHOLDER)))
-
-
-def _encode_set(o):
-    """Get a string representation of a set with its encoded content."""
-    return _encode_with_prefix(SET_PREFIX, "(%s)" % (_encode_list(list(o)).replace('"', PLACEHOLDER)))
-
-
-def _encode_dict(o):
-    """Get a string representation of a dict with its encoded content."""
-    return "{%s}" % (", ".join([f"{encode(k)}: {encode(v)}" for k, v in o.items()]))
-
-
-_encoder_table = {
-    str: _encode_str,
-    bool: _encode_bool,
-    int: _encode_int,
-    float: _encode_float,
-    dict: _encode_dict,
-    collections.OrderedDict: _encode_dict,
-    list: _encode_list,
-    set: _encode_set,
-    tuple: _encode_tuple,
-    type(None): _encode_none,
-}
+    def get_encoder_function(self, o) -> Callable:
+        return {
+            str: self._encode_str,
+            bool: self._encode_bool,
+            int: self._encode_int,
+            float: self._encode_float,
+            dict: self._encode_dict,
+            collections.OrderedDict: self._encode_dict,
+            list: self._encode_list,
+            set: self._encode_set,
+            tuple: self._encode_tuple,
+            type(None): self._encode_none,
+        }[o]

--- a/default/python/stub_generator/stub_generator/rtype/test_rtype_conversion.py
+++ b/default/python/stub_generator/stub_generator/rtype/test_rtype_conversion.py
@@ -11,14 +11,13 @@ from stub_generator.stub_generator.rtype.utils import get_name_for_mapping
     [
         (int, "int"),
         (Sequence[int], "Sequence[int]"),
-        (Sequence[PlanetId], "Sequence[PlanetId]"),
+        (Sequence[PlanetId], "Sequence[PlanetId[]]"),
         (Set[int], "Set[int]"),
-        (Set[PlanetId], "Set[PlanetId]"),
-        (Mapping[int, PlanetId], "Mapping[int, PlanetId]"),
+        (Set[PlanetId], "Set[PlanetId[]]"),
+        (Mapping[int, PlanetId], "Mapping[int, PlanetId[]]"),
         (Sequence[Sequence[int]], "Sequence[Sequence[int]]"),
-        (Sequence[Sequence[PlanetId]], "Sequence[Sequence[PlanetId]]"),
+        (Sequence[Sequence[PlanetId]], "Sequence[Sequence[PlanetId[]]]"),
     ],
 )
 def test_get_name_for_mapping(type_, string):
-    print()
     assert get_name_for_mapping(type_) == string

--- a/default/python/stub_generator/stub_generator/rtype/utils.py
+++ b/default/python/stub_generator/stub_generator/rtype/utils.py
@@ -1,3 +1,5 @@
+import typing
+
 _basic_types = {str, int, float, bool}
 
 
@@ -14,5 +16,5 @@ def get_name_for_mapping(annotation) -> str:
     elif isinstance(annotation, str):
         return annotation
     else:
-        args = [get_name_for_mapping(x) for x in annotation.__args__]
+        args = [str(get_name_for_mapping(x)) for x in typing.get_args(annotation)]
         return f"{annotation.__name__}[{', '.join(args)}]"

--- a/default/python/stub_generator/stub_generator/test_resource_builder.py
+++ b/default/python/stub_generator/stub_generator/test_resource_builder.py
@@ -1,7 +1,6 @@
 from io import StringIO
 
-from interface_inspector.class_processor import ClassInfo
-
+from stub_generator.interface_inspector.class_processor import ClassInfo
 from stub_generator.stub_generator.class_generator import process_instantiated_class
 from stub_generator.stub_generator.result_builder import Import, ResultBuilder, merge_imports, sort_imports
 
@@ -97,8 +96,8 @@ def test_process_instantiated_class():
     expected = "\n".join(
         [
             "class diplomaticMessage(NamedTuple):",
-            "    recipient: PlayerId",
-            "    sender: PlayerId",
+            "    recipient: PlayerId[]",
+            "    sender: PlayerId[]",
             "    type: diplomaticMessageType",
         ]
     )

--- a/default/python/tests/AI/save_game_codec/test_savegame_manager.py
+++ b/default/python/tests/AI/save_game_codec/test_savegame_manager.py
@@ -2,6 +2,7 @@ import pytest
 
 import savegame_codec
 from CombatRatingsAI import ShipCombatStats
+from savegame_codec._definitions import TrustedClasses
 
 
 class DummyTestClass:
@@ -20,35 +21,6 @@ class DummyTestClass:
 
     def dummy_class_function(self):
         pass
-
-
-@pytest.fixture()
-def trusted_scope(monkeypatch):
-    test_scope = dict(savegame_codec._decoder.trusted_classes)
-    test_scope.update(
-        {
-            __name__ + ".GetStateTester": GetStateTester,
-            __name__ + ".SetStateTester": SetStateTester,
-        }
-    )
-    monkeypatch.setattr(savegame_codec._decoder, "trusted_classes", test_scope)
-    monkeypatch.setattr(savegame_codec._definitions, "trusted_classes", test_scope)
-    monkeypatch.setattr(savegame_codec._encoder, "trusted_classes", test_scope)
-    return test_scope
-
-
-class Success(Exception):
-    pass
-
-
-class GetStateTester:
-    def __getstate__(self):
-        raise Success
-
-
-class SetStateTester:
-    def __setstate__(self, state):
-        raise Success
 
 
 @pytest.fixture(
@@ -81,23 +53,25 @@ def simple_object(
     return request.param
 
 
-def test_class_with_protected_attribute(trusted_scope):
+def test_class_with_protected_attribute():
     foo = ShipCombatStats()
-    retval = savegame_codec.encode(foo)
+    trusted_classes = TrustedClasses()
+    retval = savegame_codec.encode(foo, TrustedClasses())
     assert retval
     assert isinstance(retval, str)
 
-    restored_obj = savegame_codec.decode(retval)
+    restored_obj = savegame_codec.decode(retval, trusted_classes)
     assert type(restored_obj) is type(foo)
     assert restored_obj == foo
 
 
 def check_encoding(obj):
-    retval = savegame_codec.encode(obj)
+    trusted_classes = TrustedClasses()
+    retval = savegame_codec.encode(obj, trusted_classes)
     assert retval
     assert isinstance(retval, str)
 
-    restored_obj = savegame_codec.decode(retval)
+    restored_obj = savegame_codec.decode(retval, trusted_classes)
     assert type(restored_obj) is type(obj)
     assert restored_obj == obj
 
@@ -122,21 +96,23 @@ def test_encoding_type():
         pytest.fail("Could save untrusted class")
 
 
-def test_class_encoding(trusted_scope):
+def test_class_encoding():
+    trusted_scope = TrustedClasses()
     obj = DummyTestClass()
     with pytest.raises(
         savegame_codec.CanNotSaveGameException, match="Class test_savegame_manager.DummyTestClass is not trusted"
     ):
-        savegame_codec.encode(obj)
+        savegame_codec.encode(obj, trusted_scope)
         pytest.fail("Could save game even though test module should not be trusted")
 
-    trusted_scope[__name__ + ".DummyTestClass"] = DummyTestClass
-    retval = savegame_codec.encode(obj)
+    new_trusted_scope = TrustedClasses({__name__ + ".DummyTestClass": DummyTestClass})
+
+    retval = savegame_codec.encode(obj, new_trusted_scope)
 
     assert retval
     assert isinstance(retval, str)
 
-    loaded_obj = savegame_codec.decode(retval)
+    loaded_obj = savegame_codec.decode(retval, new_trusted_scope)
     assert isinstance(loaded_obj, DummyTestClass)
     assert type(loaded_obj) is type(obj)
 
@@ -145,33 +121,11 @@ def test_class_encoding(trusted_scope):
 
     assert loaded_dict == original_dict
 
-    del trusted_scope[__name__ + ".DummyTestClass"]
-
-    with pytest.raises(
-        savegame_codec.InvalidSaveGameException,
-        match="DANGER DANGER - test_savegame_manager.DummyTestClass not trusted",
-    ):
-        savegame_codec.decode(retval)
-        pytest.fail("Could load object from untrusted module")
-
-
-@pytest.mark.usefixtures("trusted_scope")
-def test_getstate_call():
-    with pytest.raises(Success):
-        savegame_codec.encode(GetStateTester())
-        pytest.fail("__getstate__ was not called during encoding.")
-
-
-@pytest.mark.usefixtures("trusted_scope")
-def test_setstate_call():
-    with pytest.raises(Success):
-        savegame_codec.decode(savegame_codec.encode(SetStateTester()))
-        pytest.fail("__setstate__ was not called during decoding.")
-
 
 def test_enums():
     import EnumsAI
 
     test_obj = {EnumsAI.MissionType.COLONISATION: EnumsAI.MissionType.INVASION}
-    restored_obj = savegame_codec.decode(savegame_codec.encode(test_obj))
+    trusted_classes = TrustedClasses()
+    restored_obj = savegame_codec.decode(savegame_codec.encode(test_obj, TrustedClasses()), trusted_classes)
     assert test_obj == restored_obj

